p2p/raintree/peers_manager.go

```diff
@@ -0,0 +1,182 @@
+package raintree
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/pokt-network/pocket/p2p/types"
+	typesP2P "github.com/pokt-network/pocket/p2p/types"
+	cryptoPocket "github.com/pokt-network/pocket/shared/crypto"
+)
+
+// peersManager is in charge of handling operations on peers (like adding/removing them) within an AddrBook
+type peersManager struct {
+	selfAddr cryptoPocket.Address
+	eventCh  chan addressBookEvent
+
+	m  sync.RWMutex
+	wg sync.WaitGroup
+
+	addrBook typesP2P.AddrBook
+
+	addrBookMap  typesP2P.AddrBookMap
+	addrList     typesP2P.AddrList
+	maxNumLevels uint32
+}
+
+func newPeersManager(selfAddr cryptoPocket.Address, addrBook typesP2P.AddrBook) (*peersManager, error) {
+	pm := &peersManager{
+		selfAddr:     selfAddr,
+		addrBook:     addrBook,
+		eventCh:      make(chan addressBookEvent, 1),
+		addrBookMap:  make(typesP2P.AddrBookMap),
+		addrList:     make([]string, 0),
+		maxNumLevels: 0,
+	}
+
+	// inizializing map and list
+	pm.addrBookMap = make(map[string]*typesP2P.NetworkPeer, len(addrBook))
+	pm.addrList = make([]string, len(addrBook))
+	for i, peer := range addrBook {
+		addr := peer.Address.String()
+		pm.addrList[i] = addr
+		pm.addrBookMap[addr] = peer
+	}
+
+	sort.Strings(pm.addrList)
+
+	i := sort.SearchStrings(pm.addrList, pm.selfAddr.String())
```

Comment:
Why are you specifically trying to avoid using SearchString?

SearchString performs a binary search and it leverages the fact that the slice we are searching in is sorted.

![image](https://user-images.githubusercontent.com/29378614/193922807-f1b57986-3237-482f-a5c6-30a3cfadfde3.png)

Reply:
Originally I was just avoiding a secondary linear pass through the addressbook by leveraging the first forloop, but I naively assumed `SearchString` is linear. LGTM